tests/config.py

```diff
@@ -2,5 +2,5 @@
 pcicTcpPort = 50010
 maxNumberContainers = 9
 # Set this to False if the applications for the unittests are already imported on the sensor
-importDeviceConfigUnittests = False
+importDeviceConfigUnittests = True
```

Comment:
IP will be changed to default IP 192.168.0.69
Flag importDeviceConfigUnittests willbe set to True

Reply:
_I can't find a need for external device in Unit Tests, that's for system testing._
How do you mean that? Unittests can only run with external device connected.

Reply:
Unit tests shall not run with external devices.... Shall not connect. Shall not use real "objects" that are not part of the unit. You don't have full control over camera. You don't know which FW it is using. You can't throw specific exceptions.

We've tested whole tcp-pcic application using unit tests that never connected to anything and we can run them from PC without a problem. They are executed with every push to the git repo to check if we didn't make the app bad.

In python you can do unit testing extremely easy, because you can mock everything, every time, everywhere.

Reply:
I understand. I see that an issue #25 exists for that topic. The merge is urgent because the customer is waiting for the new features...

Reply:
Yep there is an issue added. This PR adds to this issue even more. Additional question. What is the code coverage and branch coverage for tests on new features presented in this PR?

Reply:
I was running the coverage package on all unittests with following code coverage:

PS C:\git\o2x5xx-python> coverage report -m               
Name                              Stmts   Miss  Cover   Missing
---------------------------------------------------------------
source\__init__.py                    4      0   100%
source\device\__init__.py             3      0   100%
source\device\client.py              54      0   100%
source\device\discovery.py           53     44    17%   11-22, 26-37, 40-91
source\device\image_client.py        64     45    30%   15-33, 42, 51-56, 66-90, 100-105, 120-128
source\pcic\__init__.py               1      0   100%
source\pcic\client.py               214      8    96%   100, 278-279, 415, 643-644, 657-658
source\rpc\__init__.py                1      0   100%
source\rpc\application.py           224     29    87%   23-33, 219, 235, 342-347, 364-369, 395, 402, 428, 435, 519-520, 540
source\rpc\client.py                147     23    84%   46-47, 69, 88-89, 105-108, 173-177, 262-263, 295-301, 311
source\rpc\edit.py                   40     10    75%   19-24, 34, 70, 73, 100
source\rpc\imageQualityCheck.py     139     31    78%   57, 71, 98, 112, 141, 155, 184, 198, 211, 227-229, 241-263
source\rpc\imager.py                184     18    90%   177-180, 195-199, 203, 362-364, 392-394, 409, 431-433, 454
source\rpc\proxy.py                 121     14    88%   26-27, 31-32, 36-37, 98-99, 123, 143-147, 225
source\rpc\session.py               113     20    82%   26-31, 39-41, 53, 67, 87, 91, 106, 121, 133, 147, 209, 248, 258
source\rpc\utils.py                  33      3    91%   36, 41-43
source\static\__init__.py             2      0   100%
source\static\configs.py              1      0   100%
source\static\devices.py             19      2    89%   52-53
source\static\formats.py              4      0   100%
tests\__init__.py                     0      0   100%
tests\config.py                       4      0   100%
tests\test_application.py           256     10    96%   20-24, 30-35
tests\test_device.py                181     24    87%   22-26, 32-37, 76, 81, 126-130, 136-141, 180, 185
tests\test_edit.py                   94     10    89%   19-23, 29-34
tests\test_imageQualityCheck.py     115     26    77%   19-23, 29-34, 117-133
tests\test_imager.py                220     19    91%   19-23, 29-34, 131-140
tests\test_pcic.py                  328     14    96%   27-31, 37-42, 83, 107, 116, 316
tests\test_rpc.py                   151     18    88%   21-25, 31-36, 51-52, 104-107, 111, 117
tests\test_session.py               133     10    92%   19-23, 29-34
tests\utils.py                       14      5    64%   7-8, 15-19
---------------------------------------------------------------
TOTAL                              2917    383    87%

Reply:
And the branch coverage:

Name                              Stmts   Miss Branch BrPart  Cover

source\__init__.py                    4      0      0      0   100%
source\device\__init__.py             3      0      0      0   100%
source\device\client.py              54      0     16      0   100%
source\device\discovery.py           53     44     14      0    13%
source\device\image_client.py        64     45     14      0    24%
source\pcic\__init__.py               1      0      0      0   100%
source\pcic\client.py               214      8     40     13    92%
source\rpc\__init__.py                1      0      0      0   100%
source\rpc\application.py           224     29     52      9    83%
source\rpc\client.py                147     22     16      6    80%
source\rpc\edit.py                   40     10     10      3    74%
source\rpc\imageQualityCheck.py     139     31     52      9    73%
source\rpc\imager.py                184     18     50      6    87%
source\rpc\proxy.py                 121     13     12      3    85%
source\rpc\session.py               113     22     33     13    73%
source\rpc\utils.py                  33      3      6      2    87%
source\static\__init__.py             2      0      0      0   100%
source\static\configs.py              1      0      0      0   100%
source\static\devices.py             19      2     10      4    72%
source\static\formats.py              4      0      0      0   100%
tests\__init__.py                     0      0      0      0   100%
tests\config.py                       4      0      0      0   100%
tests\test_application.py           256     10      6      2    95%
tests\test_device.py                181     24     16      4    84%
tests\test_edit.py                   94     10     26      2    88%
tests\test_imageQualityCheck.py     115     26      6      2    75%
tests\test_imager.py                220     19      8      3    89%
tests\test_pcic.py                  328     12     34      5    95%
tests\test_rpc.py                   158     48     18      6    65%
tests\test_session.py               133     27     10      2    78%
tests\utils.py                       14      5      6      2    55%

TOTAL                              2924    428    455     96    82%

Reply:
Ok i think i understood your message now. Sorry for the long way to that...
I think we should first setup the socket mock and after that we should add tests for a full code and branch cover of especially the proxy and clients!?!?

Reply:
Yes if that is possible. Btw. not only socket, everything that is not related to the tested unit. So we have control.

Reply:
As discussed with @cfreundl we will change that in a few weeks #25 . Customer needs the new features now.